cloudflare.go

```diff
@@ -270,19 +278,102 @@ func (api *API) makeRequestWithAuthTypeAndHeaders(ctx context.Context, method, u
 			return nil, errors.Errorf("HTTP status %d: service failure", resp.StatusCode)
 		}
 
-		errBody := &Response{}
-		err = json.Unmarshal(respBody, &errBody)
-		if err != nil {
-			return nil, errors.Wrap(err, errUnmarshalErrorBody)
-		}
+		// Logpull/received error response are not in json.
+		if len(respBody) > 0 && respBody[0] == '{' {
+			errBody := &Response{}
+			err = json.Unmarshal(respBody, &errBody)
+			if err != nil {
+				return nil, errors.Wrap(err, errUnmarshalErrorBody)
+			}
 
+			return nil, &APIRequestError{
+				StatusCode: resp.StatusCode,
+				Errors:     errBody.Errors,
+			}
+		}
 		return nil, &APIRequestError{
 			StatusCode: resp.StatusCode,
-			Errors:     errBody.Errors,
+			Errors:     []ResponseInfo{{Message: string(respBody)}},
+		}
+
+	}
+	return getBodyReader(resp)
+}
+
+var gzipPool sync.Pool
+
+type gzipResponseBody struct {
+	body io.ReadCloser
+	gzip *gzip.Reader
+}
+
+type closeDiscardBody struct {
+	body io.ReadCloser
+}
+
+func (b *closeDiscardBody) Read(p []byte) (n int, err error) {
+	return b.body.Read(p)
+}
+
+func (b *closeDiscardBody) Close() error {
+	_, _ = io.Copy(ioutil.Discard, b.body)
```

Comment:
Despite that here we use `ioutil.Discard` (that will succeed on any write), If `b.body.Read()` fails, the error will be returned. Should we care about this scenario?
```suggestion
	_, err := io.Copy(ioutil.Discard, b.body)
	if err != nil {
		return errors.Wrap(err, "failed to discard response body")
	}
```